### DISCUSSION TITLE
pyarmor kubernetes  support

### DISCUSSION THREAD

**Discussion by Kathir2023** [dashingsoft/pyarmor#1336] [Ideas]
What is kubernetes cluster?
   Kubernetes cluster is a collection of nodes (computers or virtual machines) that work together to run containerized applications. Kubernetes is an open-source container orchestration platform that automates the deployment, scaling, and management of containerized applications
Pyarmor support Docker?
 Yes, we are able to test docker images with multiple VM then good to move kubernetes

Idea of implementation:
   - deploy container image in kubernetes cluster and test application
   - we may get issue K8s always create dynamic nodes based on usage
   - need to discuss standard option to implement in kubernetes support pyarmor

**Comment by jondy** [comment]
Do you mean using group license with Kubernetes container?

In this case, the solution is almost same as docker only Kubernetes container could visit Kubernetes host by ipv4 in same network.

And how to tell it's Kubernetes container? For docker, there is no disk storage.

**Reply by devopsdymyr** [reply]
@jondy I don't know how the new node can use our docker image, kubernetes node group can be scale down to 0 and scale up, if you tested with any cluster please update in the documents.

**Comment by jondy** [comment]
I have a question, is it necessary to obfuscate the Python scripts in Kubernetes container? Why not obfuscate the scripts in local machine, then deploy obfuscated scripts to Kubernetes container directly?

**Reply by Kathir2023** [reply]
yes, if possible to obfuscate in local and same image can deploy in to kubernetes. then this is wonderful.

My expectation is obfuscate python code in local or in the ci-cd pipeline with group license and deploy into kubernetes.

if you have time lets discuss in the page for 10 Min

**Reply by jondy** [reply]
I don't know Kubernetes.

If you could deploy your plain Python scripts to Kubernetes, you also could first obfuscate the script in develop machine, then deploy the obfuscated script to Kubernetes, almost same as what you do with plain scripts.

**Reply by Kathir2023** [reply]
@jondy  let me try that if that worked then pyarmor is rocking for the microservice application.

**Comment by fgarcia-0** [comment]
Dear @jondy , we have same kind of problem.
we have a CI/CD pipeline on kubernetes ,using containerd as runtime (not docker) and the cluster is based on EC2 spot instances auto scale groups, meaning the hosts are triggered on demand to support the load and will never be reused.
so what do you suggest ?
We have a group license.
would a pro license solve the problem by doing online obfuscation and runnign pyarmor in container as on a normal host ?

**Comment by jondy** [comment]
pro license should be OK if no more than 100 runs in 24 hours.

I don't know your build process, but ci/cd pipeline should get python scripts somewhere.

So you can first use docker containers or any offline device to obfuscate python scripts, then upload the obfuscated scripts to same place, then ci/cd pipeline could get these obfuscated scripts directly.

It just changes a little in ci/cd pipeline.

Why must obfuscate the scripts on kubernetes?
